location-history-handler.go

```diff
@@ -0,0 +1,296 @@
+package main
+
+import (
+	"archive/zip"
+	"bufio"
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/bikedataproject/go-bike-data-lib/dbmodel"
+	"github.com/google/uuid"
+	geo "github.com/paulmach/go.geo"
+	log "github.com/sirupsen/logrus"
+)
+
+// PointActivity : Single activity information object
+type PointActivity struct {
+	Type       string `json:"type"`
+	Confidence int    `json:"confidence"`
+}
+
+// PointActivities : Collection of activities
+type PointActivities struct {
+	TimestampMs string          `json:"timestampMs"`
+	Activity    []PointActivity `json:"activity"`
+}
+
+// LocationHistoryPoint : Single location datapoint
+type LocationHistoryPoint struct {
+	TimestampMs string            `json:"timestampMs"`
+	LatitudeE7  float64           `json:"latitudeE7"`
+	LongitudeE7 float64           `json:"longitudeE7"`
+	Accuracy    int               `json:"accuracy"`
+	Activity    []PointActivities `json:"activity,omitempty"`
+}
+
+// LocationHistory : Collection of LocationHistoryPoints
+type LocationHistory struct {
+	Locations []LocationHistoryPoint `json:"locations"`
+}
+
+// HandleLocationFile : Parse a given JSON file and process it's contents
+func HandleLocationFile(filepath string, user dbmodel.User) error {
+	// Attempt to read the file
+	data, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal file
+	var history LocationHistory
+	if err = json.Unmarshal(data, &history); err != nil {
+		return fmt.Errorf("Could not unmarshall data into location history: %v", err)
+	}
+
+	if len(history.Locations) > 0 {
+		// Convert history to trip-based objects
+		trips := make(map[string][]LocationHistoryPoint)
+
+		// Loop over each individual point & organise per day
+		for _, point := range history.Locations {
+			unixMs, err := strconv.ParseInt(point.TimestampMs, 10, 64)
+			if err != nil {
+				return err
+			}
+			timestamp := time.Unix(unixMs/1000, 0)
+
+			// Loop over the activities for each point
+			for _, actCollection := range point.Activity {
+				for _, act := range actCollection.Activity {
+					if act.Type == LocationHistoryCylcingType && act.Confidence >= LocationHistoryActivityThreshold {
+						// Set trip
+						trips[timestamp.Format("2006-01-02")] = append(trips[timestamp.Format("2006-01-02")], point)
```

Comment:
Why this date?

Reply:
Great question! The date formatting Strings in Go are based on the Posix date command. If you use "1 2 3 4 5 6" as input there, you get "Jan 2 15:04:05 2006", which is the reference time used in formatting for Go.